app/flatpak-builtins-list.c

```diff
@@ -38,16 +38,104 @@ static gboolean opt_runtime;
   { "app", 0, 0, G_OPTION_ARG_NONE, &opt_app, N_("List installed applications"), NULL },
   { "arch", 0, 0, G_OPTION_ARG_STRING, &opt_arch, N_("Arch to show"), N_("ARCH") },
   { "all", 'a', 0, G_OPTION_ARG_NONE, &opt_all, N_("List all refs (including locale/debug)"), NULL },
+  { "show-columns", 0, 0, G_OPTION_ARG_NONE, &opt_show_cols, N_("Show available columns"), NULL },
+  { "columns", 0, 0, G_OPTION_ARG_STRING_ARRAY, &opt_cols, N_("What information to show"), N_("FIELD,…")  },
   { NULL }
 };
 
+static struct {
+  const char *name;
+  const char *title;
+  const char *desc;
+} all_columns[] = {
+  { "application",  N_("Application"),    N_("Show the application ID") },
+  { "arch",         N_("Architecture"),   N_("Show the architecture") },
+  { "branch",       N_("Branch"),         N_("Show the branch") },
+  { "ref",          N_("Ref"),            N_("Show the ref") },
+  { "origin",       N_("Origin"),         N_("Show the origin remote") },
+  { "active",       N_("Active commit"),  N_("Show the active commit") },
+  { "latest",       N_("Latest commit"),  N_("Show the latest commit") },
+  { "size",         N_("Installed size"), N_("Show the installed size") },
+  { "options",      N_("Options"),        N_("Show options") },
+};
+
+#define ALL_COLUMNS "application,arch,branch,origin,active,latest,size,options"
+#define DEFAULT_COLUMNS "ref,options"
+
+static char *
+column_help (void)
+{
+  GString *s = g_string_new ("");
+  int len;
+  int i;
+
+  g_string_append (s, _("Available columns:\n"));
+
+  len = 0;
+  for (i = 0; i < G_N_ELEMENTS (all_columns); i++)
+    len = MAX (len, strlen (all_columns[i].name));
+
+  len += 4;
+  for (i = 0; i < G_N_ELEMENTS (all_columns); i++)
+    g_string_append_printf (s, "  %-*s %s\n", len, all_columns[i].name, all_columns[i].desc);
+
+  g_string_append_printf (s, "  %-*s %s\n", len, "all", _("Show all columns"));
+  g_string_append_printf (s, "  %-*s %s\n", len, "help", _("Show available columns"));
+
+  return g_string_free (s, FALSE);
+}
+
+static void
+show_columns (void)
+{
+  g_autofree char *col_help = column_help ();
+  g_print ("%s", col_help);
+}
+
+static int
+find_column (const char *name,
+             GError **error)
+{
+  int i;
+  int candidate;
+
+  candidate = -1;
+  for (i = 0; i < G_N_ELEMENTS (all_columns); i++)
+    {
+      if (g_str_equal (all_columns[i].name, name))
+        {
+          return i;
+        }
+      else if (g_str_has_prefix (all_columns[i].name, name))
+        {
+          if (candidate == -1)
+            {
+              candidate = i;
+            }
+          else
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED, _("Ambiguous column: %s"), name);
+              return -1;
+            }
+        }
+    }
+
+  if (candidate >= 0)
+    return candidate;
+
+  g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED, _("Unknown column: %s"), name);
+  return -1;
+}
+
```

Comment:
This is duplicated from ps, can we put it in app/flatpak-builtins-utils.c instead?